Tooltip Text support for Views

Redth opened the pull request:
### Description of Change

### Open questions

1) ~Everyone ok with an attached property?~ Yes they are

2)  Is `ToolTipProperty` an ok name? WinUI uses `ToolTipService`.  I mainly went with ToolTipProperty because of `SemanticProperties` and `AutomationProperties`. We don't really have anything suffixed with `Service`

3)  is  `IToolTipElement` a valid name for something that has a `ToolTip`? Similar question here https://github.com/dotnet/maui/pull/9174#discussion_r944063385

4) Should `ToolTip` be a concrete class on `ToolTipProperty` and inherit from `Element` so you can set it in one go?  We have cases for both. Right now we are following more of the `Font` approach here.  https://docs.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.controls.tooltipservice?view=windows-app-sdk-1.1  . The main reason to do this is if we want to allow people to instantiate an entire ToolTip and if we think specifying via a tooltip class is easier for the user. 

### API CHANGES

#### MAUI.Controls

```C#
// Other XAML Frameworks use ToolTipService
public class ToolTipProperties 
{
	public static readonly BindableProperty TextProperty = BindableProperty.CreateAttached("Text", typeof(string), typeof(ToolTipProperties), defaultValue: null, propertyChanged: OnToolTipPropertyChanged);
}
```

#### MAUI.Core

```C#
namespace Microsoft.Maui
{
	public partial class ToolTip
	{
		/// <summary>
		/// ToolTip content.
		/// </summary>
		public object? Content { get; set; }
	}
}
```

```C#
namespace Microsoft.Maui
{
	/// <summary>
	/// Indicates that this element has a ToolTip to show
	/// </summary>
	public interface IToolTipElement
	{
		ToolTip? ToolTip { get; }
	}
}
```


### Samples

Windows:
![image](https://user-images.githubusercontent.com/271950/180060431-bd849ec5-1ab5-4b49-a011-a5a6a9283dbe.png)

MacCatalyst:
![image](https://user-images.githubusercontent.com/271950/180060481-66a0305c-20f7-4345-beba-194a7482c4cc.png)

Android:
![image](https://user-images.githubusercontent.com/271950/180060511-b358977f-becf-42ce-a330-6d2c5b89ebb5.png)


### Issues Fixed

Partially implements: https://github.com/dotnet/maui/issues/1889  There's no placement property here yet, the UIKit API's seem to allow this customization, nor do the Android API's.

Redth committed:
Add tooltip interface and property to view

Redth committed:
Declare new tooltip API's

Redth committed:
Use tooltip text in sample

Redth committed:
Switch to tooltipcompat on android

Redth committed:
Added tooltip text tests

mattleibow commented on src/Controls/tests/DeviceTests/Elements/VisualElementTests.cs:
```diff
@@ -162,5 +165,54 @@ public async Task LoadedFiresOnPushedPage()
+#endif
+		}
+
+		const string ToolTipTextStringValue = "ToolTip Text Here!";
```
Should this be in the Core tests?

Redth replied:
Moved these to core.

mattleibow reviewed (commented):
Looks very nice and simple tbh. Love it.

rachelkang reviewed (dismissed):
LGTM so far!

I think we can benefit from adding some semantics here as well. From an accessibility perspective, the tooltip behavior is inconsistent between the platforms (tested Android, Catalyst) and in my opinion, not ideal on any at the moment.

On Android:
1. When the screen reader focus is on the button, the tooltip gets announced (as a hint) in addition to the text (this is good)
2. When long pressing to activate the hover with screen reader on, the toolitip appears visually but the screen reader does not note anything (this can be improved by using the Announce API imho)

MacCatalyst:
1. The tooltip text never gets announced (should at least get announced as a hint)
2. Same as Android

^ Those are my two cents on how it should behave from a basic accessibility standpoint. There are more complexities involved when we consider things like how to dismiss the tooltip message (the "mouse off" equivalent of screen reader - would probably need to consider the various types of announcement APIs that interrupt you vs. let you finish)

cc @PureWeen

rmarinho commented on src/Core/src/Core/ITooltip.cs:
```diff
@@ -0,0 +1,17 @@
+		/// <summary>
+		/// Gets the Tooltip text value.
+		/// </summary>
+		string? TooltipText { get; }
```
Can we get a btter name?

Redth replied:
Such as?

Redth commented:
> LGTM so far!
> 
> I think we can benefit from adding some semantics here as well. From an accessibility perspective, the tooltip behavior is inconsistent between the platforms (tested Android, Catalyst) and in my opinion, not ideal on any at the moment.
> 
> On Android:
> 
> 1. When the screen reader focus is on the button, the tooltip gets announced (as a hint) in addition to the text (this is good)
> 2. When long pressing to activate the hover with screen reader on, the toolitip appears visually but the screen reader does not note anything (this can be improved by using the Announce API imho)
> 
> MacCatalyst:
> 
> 1. The tooltip text never gets announced (should at least get announced as a hint)
> 2. Same as Android
> 
> ^ Those are my two cents on how it should behave from a basic accessibility standpoint. There are more complexities involved when we consider things like how to dismiss the tooltip message (the "mouse off" equivalent of screen reader - would probably need to consider the various types of announcement APIs that interrupt you vs. let you finish)
> 
> cc @PureWeen

On android, does it make sense that since the screenreader reads the text on focus, it's redundant to read it again on long press?  I'm hesitant to change the default behavior here given the logical path to activating the tooltip through a screenreader would already read out the text before that action.

For catalyst, what event would we ideally hook into for announcing the tooltip?  focused?

I think we could maybe open a new issue to track improvements to the screenreader experience around tooltips and keep this PR focused on the API wrapping each of the platform implementations.

Redth committed:
Move tests to core

Redth committed:
Add missing test implementations

Redth committed:
Skip tooltip test on old android api levels

Tooltips are only supported in the Android API as of 26+.  Prior to this, the TooltipCompat.SetToolTipText emulates the behavior however there is no TooltipCompat.GetToolTipText equivalent, so there's no way to actually 'get' the text that's currently set For now there's no valid way to test these cases, so we'll skip them for now.

mattleibow commented:
Now I am gonna be that guy... ToolTip or Tooltip? Should the T in Tip be uppercase or lowercase?

campersau commented:
| Framework | Spelling |
| ------------- | ------------- |
| WinUI3 | [`ToolTip`](https://github.com/microsoft/WinUI-Gallery/blob/260cb720ef83b3d134bc4805cffcfac9461dce33/WinUIGallery/ControlPages/ToolTipPage.xaml) |
| WPF | [`ToolTip`](https://github.com/dotnet/wpf/blob/9a89bc76b2d13d906c31db46728e2bdd79efa946/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/tooltip.cs) |
| WinForms | [`ToolTip`](https://github.com/dotnet/winforms/blob/ca1199eb6cfaf8ecd26ee429761764ea264dd2f5/src/System.Windows.Forms/src/System/Windows/Forms/ToolTip.cs) |
| iOS | [`toolTip`](https://developer.apple.com/documentation/uikit/uicontrol/3750832-tooltip) |
| Android | [`Tooltip`](https://developer.android.com/reference/androidx/appcompat/widget/TooltipCompat) |
| Tizen | [`tooltip`](https://docs.tizen.org/application/native/guides/ui/efl/mobile/component-tooltip/) |

PureWeen commented on src/Core/src/Handlers/View/ViewHandler.cs:
```diff
@@ -54,6 +54,7 @@ public abstract partial class ViewHandler : ElementHandler, IViewHandler
 				[nameof(IToolbarElement.Toolbar)] = MapToolbar,
 #endif
 				[nameof(IView.InputTransparent)] = MapInputTransparent,
+				[nameof(IView.TooltipText)] = MapTooltipText,
```
```suggestion
				[nameof(ITooltip.TooltipText)] = MapTooltipText,
```

rmarinho commented on src/Controls/src/Core/HandlerImpl/VisualElement/VisualElement.Impl.cs:
```diff
@@ -102,6 +102,17 @@ public int ZIndex
+		public static readonly BindableProperty TooltipTextProperty =
+ 			BindableProperty.Create(nameof(TooltipText), typeof(string), typeof(VisualElement), defaultValue: null);
+
+		public string TooltipText
```
This is limiting API: We should take a `object`, do a ToString() for now.

PureWeen replied:
Done

rmarinho commented:
How do i select the placement of the ToolTip?  Colors etc?

PureWeen committed:
- shift to using tooltip class in core

PureWeen committed:
- move tooltip to attached property

PureWeen committed:
- Tizen fixes

PureWeen committed:
- fix API definitions

PureWeen commented:
> LGTM so far!
> 
> I think we can benefit from adding some semantics here as well. From an accessibility perspective, the tooltip behavior is inconsistent between the platforms (tested Android, Catalyst) and in my opinion, not ideal on any at the moment.
> 
> On Android:
> 
> 1. When the screen reader focus is on the button, the tooltip gets announced (as a hint) in addition to the text (this is good)
> 2. When long pressing to activate the hover with screen reader on, the toolitip appears visually but the screen reader does not note anything (this can be improved by using the Announce API imho)
> 
> MacCatalyst:
> 
> 1. The tooltip text never gets announced (should at least get announced as a hint)
> 2. Same as Android
> 
> ^ Those are my two cents on how it should behave from a basic accessibility standpoint. There are more complexities involved when we consider things like how to dismiss the tooltip message (the "mouse off" equivalent of screen reader - would probably need to consider the various types of announcement APIs that interrupt you vs. let you finish)
> 
> cc @PureWeen

I think for now we should open a follow up issue to assess how ToolTips are typically surfaced from an a11y standpoint. All of the code in this PR uses platform concepts of ToolTips so we're not inventing anything across any platforms.

PureWeen committed:
- IToolTipElement

PureWeen committed:
Merge branch 'main' into dev/tooltips

# Conflicts:
#	src/Core/src/PublicAPI/net-ios/PublicAPI.Unshipped.txt
#	src/Core/src/PublicAPI/net-maccatalyst/PublicAPI.Unshipped.txt

jsuarezruiz commented on src/Core/src/Primitives/ToolTip.cs:
```diff
@@ -0,0 +1,14 @@
+		/// <summary>
+		/// ToolTip content.
+		/// </summary>
+		public object? Content { get; set; }
```
It's not `string? Text` to open the API in the future to allow any content? Although, with another implementation we can achieve similar results by allowing richer content, the default native APIs expect a string.

https://developer.apple.com/documentation/uikit/uicontrol/3750832-tooltip
https://developer.android.com/reference/androidx/appcompat/widget/TooltipCompat

PureWeen replied:
I changed it to `ToolTipProperty.Text` at the controls layer. Since we only support `Text` that seems the least surprising. If we start supporting `Content` then we can just add the content property.

jsuarezruiz reviewed (commented):
I would miss something like an **IsOpen** property, but as a basic implementation it works fine on Windows and macOS.

PureWeen committed:
Merge branch 'main' into dev/tooltips

PureWeen committed:
- change content to text

PureWeen committed:
Update ToolTipProperties.cs

PureWeen committed:
- delete override file if it exists

PureWeen commented:
> I would miss something like an **IsOpen** property, but as a basic implementation it works fine on Windows and macOS.

I agree but I don't think this is super easy to achieve with the platform APIs. I haven't found a great way to detect this on `Android`.

mattleibow committed:
Merge branch 'main' into dev/tooltips
